titan-compiler/coder.lua

```diff
@@ -238,81 +338,101 @@ local function codefor(ctx, node)
     local csstats, csexp = codeexp(ctx, node.start)
     local cfstats, cfexp = codeexp(ctx, node.finish)
     local cinc = ""
-    local cstart, cfinish, cstep, ccmp
+    local cvtyp
     if types.equals(node.decl._type, types.Integer) then
-        cstart = string.format([[
-            %s
-            lua_Integer _forstart = %s;
-        ]], csstats, csexp)
-        cfinish = string.format([[
-            %s
-            lua_Integer _forlimit = %s;
-        ]], cfstats, cfexp)
+        cvtyp = "lua_Integer"
     else
-        cstart = string.format([[
-            %s
-            lua_Number _forstart = %s;
-        ]], csstats, csexp)
-        cfinish = string.format([[
-            %s
-            lua_Number _forlimit = %s;
-        ]], cfstats, cfexp)
+        cvtyp = "lua_Number"
     end
+    local cstart = output([[
+        $CSSTATS
+        $CVTYP _forstart = $CSEXP;
+    ]], {
+        CSSTATS = csstats,
+        CSEXP = csexp,
+        CVTYP = cvtyp,
+    })
+    local cfinish = output([[
+        $CFSTATS
+        $CVTYP _forlimit = $CFEXP;
+    ]], {
+        CFSTATS = cfstats,
+        CFEXP = cfexp,
+        CVTYP = cvtyp,
+    })
+    local cstep, ccmp
+    local subs = {
+        CVAR = node.decl._cvar,
+    }
     if node.inc then
         local ilit = node2literal(node.inc)
         if ilit then
             if ilit > 0 then
+                subs.ILIT = c_integer_literal(ilit)
+                local tmpl
                 if types.equals(node.decl._type, types.Integer) then
-                    cstep = node.decl._cvar .. " = l_castU2S(l_castS2U(" .. node.decl._cvar .. ") + " .. tostring(ilit) .. ")"
+                    tmpl = "$CVAR = l_castU2S(l_castS2U($CVAR) + $ILIT)"
                 else
-                    cstep = node.decl._cvar .. " += " .. tostring(ilit)
+                    tmpl = "$CVAR += $ILIT"
                 end
-                ccmp = node.decl._cvar .. " <= _forlimit"
+                cstep = output(tmpl, subs)
+                ccmp = output("$CVAR <= _forlimit", subs)
             else
+                subs.NEGILIT = c_integer_literal(-ilit)
                 if types.equals(node.decl._type, types.Integer) then
-                    cstep = node.decl._cvar .. " = l_castU2S(l_castS2U(" .. node.decl._cvar .. ") - " .. tostring(-ilit) .. ")"
+                    cstep = output("$CVAR = l_castU2S(l_castS2U($CVAR) - $NEGILIT)", subs)
                 else
-                    cstep = node.decl._cvar .. " -= " .. tostring(ilit)
+                    cstep = output("$CVAR -= $ILIT", subs)
```

Comment:
Re-reading the code I just spotted a bug (that I wrote myself)... ILIT is not defined in subs at this point. The fact that this wasn't caught by Busted means we don't have a test covering this line!

Reply:
Pushed a fix for this, and tests